### PR TITLE
Vectorize linspace on CPU.

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27957 Vectorize linspace on CPU.**

Benchmark (gcc 8.3, Debian Buster, turbo off, Release build, Intel(R) Xeon(R) E-2136):

```python
import timeit
for dtype in ('torch.double', 'torch.float', 'torch.uint8', 'torch.int8', 'torch.int16', 'torch.int32', 'torch.int64'):
    for n, t in [(40_000, 50000),
                (400_000, 5000)]:
        print(f'torch.linspace(0, 10, {n}, dtype={dtype}) for {t} times')
        print(timeit.timeit(f'torch.linspace(0, 10, {n}, dtype={dtype})', setup=f'import torch', number=t))
```

Before:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
1.3964195849839598
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
1.2374563289922662
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
1.8631796519621275
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
1.6991038109990768
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
1.8358083459897898
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
1.7214750979910605
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
1.8356257299892604
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
1.706238206999842
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
1.7463878280250356
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
1.6172360889613628
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
1.8656846070080064
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
1.714238062966615
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
1.8272205490502529
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
1.6409171230043285
```

After:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
1.0077099470072426
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
0.8227124120458029
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
1.0058343949494883
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
0.8376779520185664
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
1.903041019977536
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
1.7576498500420712
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
1.7628699769848026
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
1.6204477970022708
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
2.0970272019621916
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
1.9493417189805768
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
2.29020385700278
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
2.1212510910118
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
2.3479344319785014
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
2.156775983981788
```

Differential Revision: [D20773454](https://our.internmc.facebook.com/intern/diff/D20773454)

### CODE REVIEW

**Comment by gchanan** [comment]
@VitalyFedyunin can you review this?

**Comment by VitalyFedyunin** [comment]
in progress

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Not going to work properly if result tensor is non contiguous (for example it is transposed 2d tensor with numel == steps). 

Not going to work properly if result tensor is large (due to parallelism of `cpu_kernel_vec`, consider using serial vec). 

Please write tests for both cases.

**Comment by xuhdev** [comment]
> Not going to work properly if result tensor is non contiguous (for example it is transposed 2d tensor with numel == steps).

The test has already added a non-contiguous case. The reason it is handled is that `cpu_kernel_vec` has specifically handled it (and this is why I removed specific contiguity processing in the original implementation):

https://github.com/pytorch/pytorch/blob/174e1ba3b8ed640c09b5ff927227e765193d0cad/aten/src/ATen/native/cpu/Loops.h#L210

> 
> Not going to work properly if result tensor is large (due to parallelism of `cpu_kernel_vec`, consider using serial vec).
> 

I've now added a test case for large tensor. From my understanding, `vectorized_loop` (called from `cpu_kernel_vec`) merely loops over `Vec` objects using SIMD instructions, but does not use multi-CPU level parallelization. Therefore, I don't think this PR will break large tensors using `cpu_kernel_vec`.

**Comment by xuhdev** [comment]
@VitalyFedyunin Mind reviewing again :)?

**Comment by VitalyFedyunin** [comment]
Fails correctness check (please include as test):

```python
In [27]: y = torch.linspace(0, 1000000-1, 1000000)                                                                                                                     

In [28]: correct = True                                                                                                                                                

In [29]: for i in range(y.shape[0]-1): 
    ...:     correct = correct and y[i] < y[i+1] 
    ...:                                                                                                                                                               

In [30]: correct                                                                                                                                                       
Out[30]: tensor(False)
```

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "linspace_cpu", [&]() {
+    int64_t idx = 0;
+    scalar_t start = scalar_start.to<scalar_t>();
+    scalar_t end = scalar_end.to<scalar_t>();
+    scalar_t step = (end - start) / static_cast<scalar_t>(steps - 1);
+    cpu_kernel_vec(
+        iter,
+        [start, step, &idx]() -> scalar_t {
+          scalar_t res = start + idx * step;
+          ++ idx;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: You **can't** use `idx` in multiple threads, please calculate it based on start and step.

**Review by VitalyFedyunin — APPROVED** [review]
Concurrently and incorrectly using `idx`

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Concurrently and incorrectly using idx

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+using namespace vec256;
+
+static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar scalar_end, int64_t steps) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.dtype(), "linspace_cpu", [&]() {
+    int64_t idx(0);
+    scalar_t start = scalar_start.to<scalar_t>();
+    scalar_t end = scalar_end.to<scalar_t>();
+    scalar_t step = (end - start) / static_cast<scalar_t>(steps - 1);
+    cpu_kernel_vec(
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: You need to use serial kernel here if you want to modify `idx`.

Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: Thanks. I tried using `atomic` yesterday but it seems less efficient than the current implementation. I'll try serial_kernel later on.

Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: I switched to `cpu_serial_kernel` and updated the benchmark. I was expecting no perf difference, but surprised that the new code is magically more efficient for float.

**Comment by xuhdev** [comment]
@VitalyFedyunin This PR should be ready to review now. It turned out to be a much more complicated PR (3 PRs stacked) than it seemed to be!

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Hi! Can you please rebase stack.

**Comment by xuhdev** [comment]
@VitalyFedyunin Done!

**Comment by xuhdev** [comment]
@VitalyFedyunin All tests have passed

**Comment by xuhdev** [comment]
@VitalyFedyunin Any chance for this one? :)

**Pull request closed without merging**